services/cargohold/test/integration/API/V4.hs

```diff
@@ -0,0 +1,282 @@
+
+--------------------------------------------------------------------------------
+-- Client compatibility tests
+
+-- Since the other tests use functions from the server code, it can happen that
+-- an API change also changes the requests made here in the tests.
+-- This test tries to prevent us from breaking the API without noticing.
+--
+-- The body is taken directly from a request made by the web app
+-- (just replaced the content with a shorter one and updated the MD5 header).
+testUploadCompatibility :: TestSignature ()
+testUploadCompatibility c = do
+  uid <- liftIO $ Id <$> nextRandom
+  -- Initial upload
+  r1 <-
+    uploadRaw (c . path "/assets/v3") uid exampleMultipart
+      <!! const 201 === statusCode
+  let loc = decodeHeader "Location" r1 :: ByteString
+  -- Lookup and download via redirect.
+  r2 <-
+    get (c . path loc . zUser uid . noRedirect) <!! do
+      const 302 === statusCode
+      const Nothing === responseBody
+  r3 <- flip get' id =<< parseUrlThrow (C8.unpack (getHeader' "Location" r2))
+  liftIO $ do
+    assertEqual "status" status200 (responseStatus r3)
+    assertEqual "content-type mismatch" (Just applicationOctetStream) (getContentType r3)
+    assertEqual "user mismatch" uid (decodeHeader "x-amz-meta-user" r3)
+    assertEqual "data mismatch" (Just "test") (responseBody r3)
+  where
+    exampleMultipart :: LByteString
+    exampleMultipart =
+      "--FrontierIyj6RcVrqMcxNtMEWPsNpuPm325QsvWQ\r\n\
+      \Content-Type: application/json;charset=utf-8\r\n\
+      \Content-length: 37\r\n\
+      \\r\n\
+      \{\"public\":true,\"retention\":\"eternal\"}\r\n\
+      \--FrontierIyj6RcVrqMcxNtMEWPsNpuPm325QsvWQ\r\n\
+      \Content-Type: application/octet-stream\r\n\
+      \Content-length: 4\r\n\
+      \Content-MD5: CY9rzUYh03PK3k6DJie09g==\r\n\
+      \\r\n\
+      \test\r\n\
+      \--FrontierIyj6RcVrqMcxNtMEWPsNpuPm325QsvWQ--\r\n\
+      \\r\n"
+
+-- API Calls ------------------------------------------------------------------
+
+uploadSimple ::
+  CargoHold ->
+  UserId ->
+  AssetSettings ->
+  (MIME.Type, ByteString) ->
+  Http (Response (Maybe Lazy.ByteString))
+uploadSimple c usr sets (ct, bs) =
+  let mp = buildMultipartBody sets ct (Lazy.fromStrict bs)
+   in uploadRaw c usr (toLazyByteString mp)
+
+uploadRaw ::
+  CargoHold ->
+  UserId ->
+  Lazy.ByteString ->
+  Http (Response (Maybe Lazy.ByteString))
+uploadRaw c usr bs =
+  post $
+    c
+      . method POST
+      . zUser usr
+      . zConn "conn"
+      . content "multipart/mixed"
+      . lbytes bs
+
+deleteAsset :: CargoHold -> UserId -> Qualified AssetKey -> Http (Response (Maybe Lazy.ByteString))
+deleteAsset c u k = delete $ c . zUser u . paths ["assets", "v3", toByteString' (qUnqualified k)]
+
+-- Utilities ------------------------------------------------------------------
+
+decodeHeader :: FromByteString a => HeaderName -> Response b -> a
+decodeHeader h =
+  fromMaybe (error $ "decodeHeader: missing or invalid header: " ++ show h)
```

Comment:
I'd prefer the return type to be `Maybe a` or similar rather than invoking `error`, but perhaps this ain't that bad given it's test code.

Reply:
The idea is that if you're looking up a header in a test, you'll likely either want to assert it's there or use it as some kind of test setup and cannot reasonably resume without it.

While looking at it again, I realized that it might make sense to add a `HasCallStack` constraint, to get more meaningful backtraces in case of error.

@mdimjasevic , do you think we should change the name; communicating that this might fail? E.g. to `decodeHeaderOrFail`?

Reply:
@supersven , I agree with adding `HasCallStack`. Yes, the renaming suggestion also sounds good to me.